File: fetch-news.py
#!/usr/bin/env python3
"""
新闻抓取脚本
运行方式: python fetch-news.py
功能: 从配置的RSS源抓取新闻，合并到cached-news.json，自动去重保留最近90天
"""

import json
import requests
import xml.etree.ElementTree as ET
from datetime import datetime, timedelta
import time
import os

# 配置
NEWS_SOURCES_FILE = 'news-sources.json'
CACHE_FILE = 'cached-news.json'
MAX_ITEMS_PER_SOURCE = 10  # 每个源最多抓取多少条
MAX_TOTAL_ITEMS = 200      # 总缓存最多保留多少条
DAYS_TO_KEEP = 90          # 保留多少天内的新闻

def load_sources():
    """加载新闻源配置"""
    with open(NEWS_SOURCES_FILE, 'r', encoding='utf-8') as f:
        sources = json.load(f)
    return [s for s in sources if s.get('enabled', True)]

def fetch_rss(source):
    """从RSS源抓取新闻"""
    try:
        headers = {
            'User-Agent': 'Mozilla/5.0 (compatible; TerryGameLab/1.0)'
        }
        response = requests.get(source['url'], headers=headers, timeout=10)
        response.raise_for_status()
        
        # 解析XML
        root = ET.fromstring(response.content)
        channel = root.find('channel')
        items = channel.findall('item') if channel else []
        
        news_list = []
        for item in items[:MAX_ITEMS_PER_SOURCE]:
            title = item.find('title').text if item.find('title') is not None else ''
            link = item.find('link').text if item.find('link') is not None else ''
            description = item.find('description').text if item.find('description') is not None else ''
            pubDate = item.find('pubDate').text if item.find('pubDate') is not None else ''
            
            # 解析时间
            try:
                if pubDate:
                    # 解析RSS时间格式
                    dt = datetime.strptime(pubDate, '%a, %d %b %Y %H:%M:%S %Z')
                    timestamp = int(dt.timestamp() * 1000)
                else:
                    timestamp = int(time.time() * 1000)
            except Exception:
                timestamp = int(time.time() * 1000)
            
            # 处理分类 - 支持多个标签
            categories = source.get('categories', [])
            if not categories and 'category' in source:
                categories = [source['category']]
            
            # 清理描述
            if description:
                # 去掉HTML标签，截断
                import re
                description = re.sub(r'<[^>]*>', '', description)
                description = description[:200] + '...' if len(description) > 200 else description
            
            news_list.append({
                'title': title.strip() if title else '',
                'link': link.strip() if link else '',
                'summary': description.strip() if description else '',
                'time': timestamp,
                'source': source['name'],
                'categories': categories,
                'sourceColor': source.get('color', 'rgba(0, 255, 255, 0.3)')
            })
            
        print(f"✅ {source['name']}: 获取到 {len(news_list)} 条新闻")
        return news_list
        
    except Exception as e:
        print(f"❌ {source['name']}: 获取失败 - {str(e)}")
        return []

def merge_news(existing_news, new_news):
    """合并新闻，按URL去重，保留最新"""
    url_map = {}
    
    # 添加已有新闻
    for news in existing_news:
        if 'link' in news and news['link']:
            url_map[news['link']] = news
    
    # 添加新新闻（会覆盖相同URL，保证最新）
    for news in new_news:
        if 'link' in news and news['link']:
            url_map[news['link']] = news
    
    # 转回数组
    merged = list(url_map.values())
    
    # 按时间倒序
    merged.sort(key=lambda x: x['time'], reverse=True)
    
    # 过滤掉过期新闻
    cutoff_time = int((datetime.now() - timedelta(days=DAYS_TO_KEEP)).timestamp() * 1000)
    merged = [n for n in merged if n['time'] > cutoff_time]
    
    # 限制总数
    if len(merged) > MAX_TOTAL_ITEMS:
        merged = merged[:MAX_TOTAL_ITEMS]
    
    return merged

def main():
    print("🚀 Terry Game Lab 新闻抓取工具")
    print("-" * 40)
    
    # 加载新闻源
    sources = load_sources()
    print(f"📋 加载了 {len(sources)} 个启用的新闻源")
    
    # 加载已有缓存
    existing_news = []
    if os.path.exists(CACHE_FILE):
        with open(CACHE_FILE, 'r', encoding='utf-8') as f:
            try:
                existing_news = json.load(f)
                print(f"📂 已有缓存 {len(existing_news)} 条新闻")
            except:
                print("⚠️ 缓存文件损坏，将创建新缓存")
                existing_news = []
    
    # 抓取所有新闻源
    all_new_news = []
    for source in sources:
        news = fetch_rss(source)
        all_new_news.extend(news)
        time.sleep(1)  # 礼貌延迟
    
    print("-" * 40)
    print(f"📥 总共抓取到 {len(all_new_news)} 条新闻")
    
    # 合并
    merged = merge_news(existing_news, all_new_news)
    print(f"🔗 合并后共 {len(merged)} 条新闻（去重+过滤过期）")
    
    # 保存
    with open(CACHE_FILE, 'w', encoding='utf-8') as f:
        json.dump(merged, f, ensure_ascii=False, indent=2)
    
    print(f"💾 已保存到 {CACHE_FILE}")
    print("\n下一步: git add cached-news.json && git commit && git push")
    print("推送后网站就会显示更新后的新闻了")

if __name__ == '__main__':
    main()

File: fetch-news.js
const https = require('https');
const http = require('http');
const { parseString } = require('xml2js');
const fs = require('fs');
const url = require('url');

// 读取配置
const NEWS_SOURCES = JSON.parse(fs.readFileSync('./news-sources.json', 'utf8'))
  .filter(s => s.enabled);

console.log(`加载了 ${NEWS_SOURCES.length} 个启用的RSS源`);

const CORS_PROXIES = [
  'https://api.allorigins.win/raw?url=',
  'https://api.codetabs.com/v1/proxy?quest=',
];

function getCORSProxy() {
  return CORS_PROXIES[Math.floor(Math.random() * CORS_PROXIES.length)];
}

function fetchUrl(targetUrl) {
  return new Promise((resolve, reject) => {
    const proxyUrl = getCORSProxy() + encodeURIComponent(targetUrl);
    const parsedUrl = new URL(proxyUrl);
    
    const options = {
      hostname: parsedUrl.hostname,
      path: parsedUrl.pathname + parsedUrl.search,
      method: 'GET',
      headers: {
        'User-Agent': 'Mozilla/5.0 (compatible; RSS fetcher)'
      },
      timeout: 15000
    };

    const requester = parsedUrl.protocol === 'https:' ? https : http;
    
    const req = requester.request(options, (res) => {
      let data = '';
      res.on('data', (chunk) => data += chunk);
      res.on('end', () => resolve(data));
    });
    
    req.on('error', reject);
    req.on('timeout', () => reject(new Error('timeout')));
    req.end();
  });
}

function parseRSS(xmlText, source) {
  return new Promise((resolve, reject) => {
    parseString(xmlText, (err, result) => {
      if (err) return reject(err);
      
      let items = [];
      if (result?.rss?.channel?.[0]?.item) {
        items = result.rss.channel[0].item;
      } else if (result?.feed?.entry) {
        items = result.feed.entry;
      }
      
      if (!items || items.length === 0) {
        return resolve([]);
      }
      
      const parsed = items.slice(0, 15).map(item => {
        const title = item.title?.[0]?._ || item.title?.[0] || '';
        let link = '';
        if (item.link?.[0]?.$?.href) {
          link = item.link[0].$.href;
        } else if (item.link?.[0]) {
          link = item.link[0];
        }
        let description = '';
        if (item.description?.[0]?._) {
          description = item.description[0]._;
        } else if (item.description?.[0]) {
          description = item.description[0];
        } else if (item.summary?.[0]) {
          description = item.summary[0];
        }
        let pubDate = '';
        if (item.pubDate?.[0]) {
          pubDate = item.pubDate[0];
        } else if (item.published?.[0]) {
          pubDate = item.published[0];
        } else if (item.updated?.[0]) {
          pubDate = item.updated[0];
        }
        
        let summary = description.replace(/<[^>]*>/g, '');
        if (summary.length > 200) {
          summary = summary.slice(0, 200) + '...';
        }
        
        return {
          title: title.trim(),
          link: link.trim(),
          summary: summary.trim(),
          time: new Date(pubDate).getTime(),
          source: source.name,
          categories: source.categories,
          sourceColor: source.color || 'rgba(0, 255, 255, 0.3)'
        };
      }).filter(item => item.title && item.link && !isNaN(item.time));
      
      resolve(parsed);
    });
  });
}

async function fetchSource(source) {
  try {
    console.log(`正在抓取: ${source.name} - ${source.url}`);
    const xml = await fetchUrl(source.url);
    const items = await parseRSS(xml, source);
    console.log(`  → 获取到 ${items.length} 条新闻`);
    return items;
  } catch (e) {
    console.error(`  → 抓取失败: ${e.message}`);
    return [];
  }
}

function mergeNews(existing, newNews) {
  const urlMap = new Map();
  
  // 添加已有新闻
  existing.forEach(n => urlMap.set(n.link, n));
  
  // 添加新新闻，去重
  newNews.forEach(n => urlMap.set(n.link, n));
  
  // 转回数组，按时间排序
  let merged = Array.from(urlMap.values());
  merged.sort((a, b) => b.time - a.time);
  
  // 只保留最近90天，最多200条
  const ninetyDaysAgo = Date.now() - 90 * 24 * 60 * 60 * 1000;
  merged = merged.filter(n => n.time > ninetyDaysAgo);
  
  if (merged.length > 200) {
    merged = merged.slice(0, 200);
  }
  
  return merged;
}

function generateNewsSummaryHtml(merged) {
  // 统计分类
  const categoryMap = {
    game: '新游发布',
    company: '公司动态',
    investment: '行业资本',
    industry: '行业动态',
    policy: '政策监管'
  };
  
  const counts = {
    game: 0,
    company: 0,
    investment: 0,
    industry: 0,
    policy: 0
  };
  
  merged.forEach(n => {
    n.categories.forEach(c => {
      if (counts[c] !== undefined) {
        counts[c]++;
      }
    });
  });
  
  // 获取热门标题（只保留游戏行业相关，取最近15条）
  const hotTitles = merged
    .slice(0, 15)
    .map(n => `<a href="${n.link}" target="_blank" rel="noopener">${n.title}</a>`)
    .join('</li>\n  <li>');
  
  const now = new Date().toLocaleString('zh-CN');
  
  return `
<!-- 新闻小结 START -->
<div class="news-summary">
  <div class="summary-card">
    <h3>📊 今日资讯小结</h3>
    <div class="summary-stats">
      <div class="stat-item">
        <span class="stat-number">${merged.length}</span>
        <span class="stat-label">总资讯</span>
      </div>
      <div class="stat-item">
        <span class="stat-number">${counts.game}</span>
        <span class="stat-label">新游发布</span>
      </div>
      <div class="stat-item">
        <span class="stat-number">${counts.investment}</span>
        <span class="stat-label">行业资本</span>
      </div>
      <div class="stat-item">
        <span class="stat-number">${counts.company}</span>
        <span class="stat-label">公司动态</span>
      </div>
    </div>
    <div class="hot-titles">
      <h4>🔥 最新热门</h4>
      <ul>
        <li>${hotTitles}</li>
      </ul>
    </div>
    <div class="last-update">最后更新：${now}</div>
  </div>
</div>
<!-- 新闻小结 END -->
`;
}

async function updateNewsHtmlWithSummary(merged) {
  try {
    // 读取现有的 news.html
    let htmlContent = fs.readFileSync('./news.html', 'utf8');
    
    // 检查是否已有小结，如果有则替换
    const summaryHtml = generateNewsSummaryHtml(merged);
    
    if (htmlContent.includes('<!-- 新闻小结 START -->')) {
      // 替换现有小结
      htmlContent = htmlContent.replace(
        /<!-- 新闻小结 START -->[\s\S]*<!-- 新闻小结 END -->/,
        summaryHtml
      );
    } else {
      // 在 filters 之前插入小结
      htmlContent = htmlContent.replace(
        /        <div class="filters">/,
        `${summaryHtml}\n        <div class="filters">`
      );
    }
    
    // 写回文件
    fs.writeFileSync('./news.html', htmlContent, 'utf8');
    console.log(`✅ 已更新 news.html，添加新闻小结 (${merged.length} 条新闻)`);
    return true;
  } catch (e) {
    console.error('❌ 更新 news.html 失败:', e.message);
    return false;
  }
}

async function main() {
  // 读取现有缓存
  let existing = [];
  if (fs.existsSync('./cached-news.json')) {
    try {
      existing = JSON.parse(fs.readFileSync('./cached-news.json', 'utf8'));
      console.log(`读取现有缓存: ${existing.length} 条新闻`);
    } catch (e) {
      console.log('无现有缓存，从头开始');
    }
  }
  
  // 抓取所有源
  const allResults = [];
  for (const source of NEWS_SOURCES) {
    const items = await fetchSource(source);
    allResults.push(...items);
    // 延时，避免被封
    await new Promise(r => setTimeout(r, 2000));
  }
  
  console.log(`\n总共抓取到 ${allResults.length} 条新闻`);
  
  // 合并去重
  const merged = mergeNews(existing, allResults);
  console.log(`合并去重后: ${merged.length} 条新闻`);
  
  // 保存
  fs.writeFileSync('./cached-news.json', JSON.stringify(merged, null, 2), 'utf8');
  console.log(`\n已保存到 cached-news.json，共 ${merged.length} 条`);
  
  // 更新 news.html 添加新闻小结
  await updateNewsHtmlWithSummary(merged);
  
  if (merged.length >= 100) {
    console.log('✓ 已达到100篇目标！');
  } else {
    console.log(`✗ 还差 ${100 - merged.length} 篇达到100篇`);
  }
}

main().catch(console.error);
